### PR TITLE
Get Series Optimization

### PR DESCRIPTION
**Pull request by alanprot**
Adding jsoniter encode for `labels.Labels` the API.

This can improve the serialization speed/memory quite a bit specially for GetSeries calls that returns lots of series.

```
benchstat /tmp/old /tmp/new
name              old time/op    new time/op    delta
SeriesRespond-16     247ms ± 5%     153ms ± 2%  -38.26%  (p=0.008 n=5+5)

name              old alloc/op   new alloc/op   delta
SeriesRespond-16     123MB ± 6%      67MB ± 5%  -45.48%  (p=0.008 n=5+5)

name              old allocs/op  new allocs/op  delta
SeriesRespond-16     1.18M ± 3%     0.40M ± 0%  -66.26%  (p=0.008 n=5+5)
```

### CODE REVIEW

**Comment by alanprot** [comment]
This changing is breaking the `stringlabels` implementation for now.

If others think this is a good change i can fix that.

File: web/api/v1/api.go [hunk 1]
```diff
@@ -1954,3 +1955,28 @@ func marshalExemplarJSON(ptr unsafe.Pointer, stream *jsoniter.Stream) {
...
+//	   <labelName1>: <labelValue1>,
+//	   <labelName2>: <labelValue2>,
+//	   ...
+//	   <labelNameN>: <labelValueN>
+//	}
+func marshalLabelsJSON(ptr unsafe.Pointer, stream *jsoniter.Stream) {
+	lbls := *((*labels.Labels)(ptr))
+	stream.WriteObjectStart()
+	for i, lbl := range lbls {
+		stream.WriteObjectField(lbl.Name)
```
Comment (bboreham) [inline web/api/v1/api.go:1971]: Interested to know if this version is better than https://github.com/cortexproject/cortex/blob/f19c4016e36c80d5129520903160063bd088c31a/pkg/chunk/json_helpers.go#L42

**Review by bboreham — COMMENTED** [review]
Agreed this would be good, if implemented for the now-standard build.

Suggest creating a function taking `labels.Labels` which `marshalExemplarJSON`, `marshalSeriesJSON` and `marshalSampleJSON` can call directly, thus saving an allocation.
Then the jsoniter one which takes `unsafe.Pointer` can call that too.

**Comment by alanprot** [comment]
Closing in favor of https://github.com/prometheus/prometheus/pull/12598

Thanks

**Pull request closed without merging**